File: typescript/pxprpc/extend.ts
import { Client, Io, PxpCallable, PxpRequest, Serializer, Server } from './base'


export class RpcExtendError extends Error {
}


//Client side
export class RpcExtendClientObject {
    public constructor(public client: RpcExtendClient1, public value: number | undefined) {
    }
    public async free() {
        if (this.value != undefined) {
            let sid=this.client.allocSid();
            try{
                await this.client.baseClient.freeRef([this.value],sid)
            }finally{
                await this.client.freeSid(this.value);
            }
        }
    }
    public async asCallable():Promise<RpcExtendClientCallable>{
        return new RpcExtendClientCallable(this.client,this.value)
    }
}

export class RpcExtendClientCallable extends RpcExtendClientObject {
    public tParam: string=''
    public tResult:string=''
    public constructor(client: RpcExtendClient1, value: number | undefined) {
        super(client, value)
    }
    /*
function typedecl
format: 'parameters type->return type' 
eg:
a function defined in c:
bool fn(uin32_t,uint64_t,float64_t,struct pxprpc_object *)
defined in java:
boolean fn(int,int,double,Object)
...
it's pxprpc typedecl: 
iido->c

available type typedecl characters:
i  int(32bit integer)
l  long(64bit integer) 
f  float(32bit float)
d  double(64bit float)
o  object(32bit reference address)
b  bytes(bytes buffer)
'' return void(32bit 0)

c  boolean(pxprpc use 1byte(1/0) to store a boolean value)
s  string(bytes will be decode to string)
*/
    public typedecl(decl: string) {
        [this.tParam,this.tResult]=decl.split('->');
        return this;
    }

    public async call(...args:any[]) {
        let buf:Uint8Array[]=[]
        if(this.tParam==='b'){
            let abuf=args[0] as Uint8Array
            buf=[abuf];
        }else{
            let ser=new Serializer().prepareSerializing(32);
            new TableSerializer().
                bindContext(null,this.client).bindSerializer(ser).setColumnInfo(this.tParam,null).
                putRowsData([args])
            let serbuf=ser.build()
            buf=[serbuf];
        }
        let sid=this.client.allocSid();
        try{
            this.client.throwIfNotRunning();
            let resp=await this.client.baseClient.call(this.value!,buf,sid);
            if(this.tResult==='b'){
                return resp;
            }else{
                let ser=new Serializer().prepareUnserializing(resp);
                let rets=new TableSerializer().
                    bindContext(null,this.client).bindSerializer(ser).setColumnInfo(this.tResult,null).
                    getRowsData(1)[0]
                if(this.tResult.length===1){
                    return rets[0];
                }else if(this.tResult.length===0){
                    return null;
                }else{
                    return rets;
                }
            }
        }finally{
            this.client.freeSid(sid);
        }
    }
}

export class RpcExtendClient1 {
    private __usedSid: { [index: number]: boolean | undefined } = {};
    private __nextSid: number;

    private __sidStart: number = 1;
    private __sidEnd: number = 0xffff;

    public constructor(public baseClient: Client) {
        this.__nextSid = this.__sidStart;
    }
    public serverName?:string
    public async init():Promise<this>{
        this.baseClient.run();
        for(let item of (await this.baseClient.getInfo()).split('\n')){
            if(item.indexOf(':')>=0){
                let [key,val]=item.split(':')
                if(key==='server name'){
                    this.serverName=val;
                }
            }
        }
        return this;
    }
    
    public allocSid() {
        let reachEnd = false;
        while (this.__usedSid[this.__nextSid]) {
            this.__nextSid += 1
            if (this.__nextSid >= this.__sidEnd) {
                if (reachEnd) {
                    throw new RpcExtendError('No sid available')
                } else {
                    reachEnd = true
                    this.__nextSid = this.__sidStart
                }
            }
        }

        let t1 = this.__nextSid
        this.__nextSid += 1
        if(this.__nextSid>=this.__sidEnd){
            this.__nextSid=this.__sidStart;
        }
        this.__usedSid[t1] = true;
        return t1
    }
    public freeSid(index: number) {
        delete this.__usedSid[index]
    }
    public throwIfNotRunning(){
        if(!this.baseClient.isRunning()){
            throw new RpcExtendError('baseClient is not running.');
        }
    }
    public async getFunc(name: string) {
        this.throwIfNotRunning()
        let sid=this.allocSid();
        try{
            let index=await this.baseClient.getFunc(name,sid);
            if(index===-1)return null;
            return new RpcExtendClientCallable(this, index)
        }finally{
            this.freeSid(sid);
        }
    }
    public async close(){
        await this.baseClient.close()
    }

}


//Server side
export function allocRefFor(serv:Server,obj:any){
    let ref=serv.allocRef();
    ref.object=obj;
    if((typeof obj==='object') && ('close' in obj)){
        ref.onFree=()=>obj.close()
    }
    return ref;
}

export class RpcExtendServerCallable implements PxpCallable{
    protected tParam:string = '';
    protected tResult:string='';
    public constructor(public wrapped:(...args:any)=>Promise<any>){
    }
    //See RpcExtendClientCallable.typedecl
    public typedecl(decl: string) {
        let [tParam,tResult]=decl.split('->');
        this.tParam=tParam;
        this.tResult=tResult;
        return this;
    }
    public readParameter (req: PxpRequest){
        let buf=req.parameter!
        let param=[];
        if(this.tParam==='b'){
            param=[new Uint8Array(buf.buffer,buf.byteOffset,buf.byteLength)];
        }else{
            let ser=new Serializer().prepareUnserializing(buf);
            param=new TableSerializer().
                bindContext(req.context,null).bindSerializer(ser).setColumnInfo(this.tParam,null).
                getRowsData(1)[0]
        }
        return param;
    }
    public async call(req: PxpRequest) : Promise<any>{
        try{
            let r=await this.wrapped.apply(this,this.readParameter(req));
            req.result=await this.writeResult(req,r);
        }catch(e){
            return req.rejected=e as Error;
        }
    }
    public writeResult(req: PxpRequest,r:any):Uint8Array[]{
        if(this.tResult==='b'){
            return [r]
        }else{
            let ser=new Serializer().prepareSerializing(8);
            let results=this.tResult.length>1?r:[r]
            new TableSerializer().
                bindContext(req.context,null).bindSerializer(ser).setColumnInfo(this.tResult,null).
                putRowsData([results]);
            let buf=ser.build();
            if(buf.byteLength==0)return [];
            return [buf];
        }
    }
}

export var defaultFuncMap={} as {[k:string]:PxpCallable};

defaultFuncMap['builtin.anyToString']=new RpcExtendServerCallable(async (obj:any)=>String(obj)).typedecl('o->s');
defaultFuncMap['builtin.jsExec']=new RpcExtendServerCallable(async (code:string,arg:any)=>{
    let r=(new Function('arg',code))(arg) as any;
    if(r instanceof Promise){
        r=await r;
    }
    return r;
}).typedecl('so->o'); 
defaultFuncMap['builtin.typeof']=new RpcExtendServerCallable(async (arg:any)=>typeof arg).typedecl('o->s');
defaultFuncMap['builtin.toJSON']=new RpcExtendServerCallable(async (arg:any)=>JSON.stringify(arg)).typedecl('o->s');
defaultFuncMap['builtin.fromJSON']=new RpcExtendServerCallable(async (arg:string)=>JSON.parse(arg)).typedecl('s->o');
defaultFuncMap['builtin.bufferData']=new RpcExtendServerCallable(async (arg:Uint8Array)=>arg).typedecl('o->b');
defaultFuncMap['pxprpc_pp.io_send']=new RpcExtendServerCallable(async (io:Io,arg:Uint8Array)=>io.send([arg])).typedecl('ob->');
defaultFuncMap['pxprpc_pp.io_receive']=new RpcExtendServerCallable(async (io:Io)=>await io.receive()).typedecl('o->b');

export class RpcExtendServer1{
    public constructor(public serv:Server){
        serv.funcMap=(name)=>this.findFunc(name);
    }
    public async serve(){
        await this.serv.serve()
    }
    public funcMap=defaultFuncMap;
    public findFunc(name:string){
        return this.funcMap[name]
    }
}

export class TableSerializer {
    FLAG_NO_COLUMN_NAME=1;
    columnName:string[]|null=null;
    columnType:string|null=null;
    rows:any[][]=[];
    boundServContext:Server|null=null;
    boundClieContext:RpcExtendClient1|null=null;
    ser:Serializer|null=null;
    public setColumnInfo(types:string|null,names:string[]|null){
        this.columnName=names;
        this.columnType=types;
        return this;
    }
    public bindContext(serv:Server|null,clie:RpcExtendClient1|null){
        this.boundServContext=serv;
        this.boundClieContext=clie;
        return this;
    }
    public bindSerializer(ser:Serializer){
        this.ser=ser;
        return this;
    }
    public getRow(index:number){
        return this.rows[index];
    }
    public getRowCount() {
        return this.rows.length
    }
    public addRow(row:any[]){
        this.rows.push(row);return this;
    }
    public getRowsData(rowCnt:number):any[][]{
        let rows:any[][]=[];
        let colCnt=this.columnType!.length;
        let ser=this.ser!;
        for(let i1=0;i1<rowCnt;i1++){
            rows.push(new Array(colCnt));
        }
        for(let i1=0;i1<colCnt;i1++){
            let type=this.columnType!.charAt(i1);
            switch(type){
                case 'i':
                    for(let i2=0;i2<rowCnt;i2++){
                        rows[i2][i1]=ser.getInt();
                    }
                    break;
                case 'l':
                    for(let i2=0;i2<rowCnt;i2++){ 
                        rows[i2][i1]=ser.getLong();
                    }
                    break;
                case 'f':
                    for(let i2=0;i2<rowCnt;i2++){
                        rows[i2][i1]=ser.getFloat();
                    }
                    break;
                case 'd':
                    for(let i2=0;i2<rowCnt;i2++){
                        rows[i2][i1]=ser.getDouble();
                    }
                    break;
                case 'b':
                    for(let i2=0;i2<rowCnt;i2++){
                        rows[i2][i1]=ser.getBytes();
                    }
                    break;
                case 's':
                    for(let i2=0;i2<rowCnt;i2++){
                        rows[i2][i1]=ser.getString();
                    }
                    break;
                case 'c':
                    for(let i2=0;i2<rowCnt;i2++){
                        rows[i2][i1]=ser.getVarint()!==0;
                    }
                    break;
                case 'o':
                    for(let i2=0;i2<rowCnt;i2++){
                        let val=ser.getInt();
                        if(val===-1){
                            rows[i2][i1]=null;
                        }else{
                            if(this.boundServContext!=null){
                                rows[i2][i1]=this.boundServContext.getRef(val).object;
                            }else{
                                rows[i2][i1]=new RpcExtendClientObject(this.boundClieContext!,val);
                            }
                        }
                    }
                    break;
                default:
                    throw new RpcExtendError("Unknown Type");
            }
        }
        return rows;
    }
    public load(buf:Uint8Array){
        if(buf!=null){
            this.bindSerializer(new Serializer().prepareUnserializing(buf))
        }
        let ser=this.ser!;
        let flag=ser.getVarint();
        let rowCnt=ser.getVarint();
        this.columnType=ser.getString();
        let colCnt=this.columnType.length;
        if((flag & this.FLAG_NO_COLUMN_NAME)===0){
            this.columnName=[]
            for(let i1=0;i1<colCnt;i1++){
                this.columnName.push(ser.getString());
            }
        }
        this.rows=this.getRowsData(rowCnt);
        return this;
    }
    public putRowsData(rows:any[][]){
        let colCnt=this.columnType!.length;
        let rowCnt=rows.length;
        let ser=this.ser!;
        for(let i1=0;i1<colCnt;i1++){
            let type=this.columnType!.charAt(i1);
            switch(type){
                case 'i':
                    for(let i2=0;i2<rowCnt;i2++){
                        ser.putInt(rows[i2][i1]);
                    }
                    break;
                case 'l':
                    for(let i2=0;i2<rowCnt;i2++){
                        ser.putLong(rows[i2][i1]);
                    }
                    break;
                case 'f':
                    for(let i2=0;i2<rowCnt;i2++){
                        ser.putFloat(rows[i2][i1]);
                    }
                    break;
                case 'd':
                    for(let i2=0;i2<rowCnt;i2++){
                        ser.putDouble(rows[i2][i1]);
                    }
                    break;
                case 'b':
                    for(let i2=0;i2<rowCnt;i2++){
                        ser.putBytes(rows[i2][i1]);
                    }
                    break;
                case 's':
                    for(let i2=0;i2<rowCnt;i2++){
                        ser.putString(rows[i2][i1]);
                    }
                    break;
                case 'c':
                    for(let i2=0;i2<rowCnt;i2++){
                        ser.putVarint(rows[i2][i1]?1:0);
                    }
                    break;
                case 'o':
                    for(let i2=0;i2<rowCnt;i2++){
                        let val=rows[i2][i1];
                        if(val!==null){
                            if(this.boundServContext!=null){
                                ser.putInt(allocRefFor(this.boundServContext,val).index);
                            }else{
                                ser.putInt(val.value);
                            }
                        }else{
                            ser.putInt(-1);
                        }
                    }
                    break;
                default:
                    throw new RpcExtendError("Unknown Type");
            }
        }
    }
    public build():Uint8Array{
        if(this.ser==null){
            this.bindSerializer(new Serializer().prepareSerializing(64));
        }
        let ser=this.ser!;
        if(this.columnType==null){
            this.columnType=''
            if(this.rows.length>=1){
                for(let t1 of this.rows[0]){
                    switch(typeof t1){
                        case 'number':
                            this.columnType+='d';
                            break;
                        case 'string':
                            this.columnType+='s';
                            break;
                        case 'boolean':
                            this.columnType+='c';
                            break;
                        case 'bigint':
                            this.columnType+='l';
                            break;
                        default:
                            if(t1 instanceof Uint8Array){
                                this.columnType+='b'
                            }else{
                                this.columnType+='o'
                            }
                            break;
                    }
                }
            }
        }
        let flag=0;
        if(this.columnName==null){
            flag|=this.FLAG_NO_COLUMN_NAME;
        }
        ser.putVarint(flag);
        let rowCnt=this.rows.length;
        ser.putVarint(rowCnt);
        ser.putString(this.columnType!);
        if(this.columnName!==null){
            for(let e of this.columnName){
                ser.putString(e);
            }
        }
        this.putRowsData(this.rows);
        return ser.build();
    }
    public toMapArray():any[]{
        let r:any[]=[]
        let rowCount=this.getRowCount()
        let colCount=this.columnName!.length;
        for(let t1=0;t1<rowCount;t1++){
            let r0={} as any;
            let row=this.getRow(t1);
            for(let t2=0;t2<colCount;t2++){
                r0[this.columnName![t2]]=row[t2];
            }
            r.push(r0);
        }
        return r
    }
    public fromMapArray(val:any[]):this{
        if(val.length>0 && this.columnName===null){
            this.columnName=[];
            for(let k in val[0]){
                this.columnName.push(k);
            }
        }
        let rowCount=val.length;
        let colCount=this.columnName!.length;
        for(let t1=0;t1<rowCount;t1++){
            let row=[];
            for(let t2=0;t2<colCount;t2++){
                row.push(val[t1][this.columnName![t2]]);
            }
            this.addRow(row);
        }
        return this
    }
    public toArray():any[]{
        let r:any[]=[]
        let rowCount=this.getRowCount()
        for(let t1=0;t1<rowCount;t1++){
            r.push(this.getRow(t1)[0]);
        }
        return r 
    }
    public fromArray(val:any[]){
        let rowCount=val.length;
        for(let t1=0;t1<rowCount;t1++){
            this.addRow([val[t1]]);
        }
        return this
    }
}

File: python/pxprpc/extend.py


from typing import Optional,Any,Callable,List,cast,TypeVar,Dict,Tuple,Set,Union,Awaitable
import logging
log1=logging.getLogger(__name__)
import asyncio
from .base import NotNone,PxpCallable,Serializer,PxpRequest,PxpRef,ServerContext,ClientContext
import struct
import inspect


def pytypeToDeclChar(t):
    typemap={
                int:'l',float:'d',bool:'c',bytes:'b',str:'s'
            }
    return typemap.get(t,'o')

class TableSerializer:
    FLAG_NO_HEADER_NAME=1
    def __init__(self):
        self.rows:List[List[Any]]=[]
        self.headerType=None
        self.headerName=None
        self.ser=None
        self.boundServContext=None
        self.boundClieContext=None

    def setHeader(self,types:Optional[str],names:Optional[List[str]]):
        self.headerName=names
        self.headerType=types
        return self
    
    def getRow(self,index:int):
        return self.rows[index]
    
    def getRowCount(self):
        return len(self.rows)
    
    def addRow(self,row:List[Any]):
        self.rows.append(row)
        return self
    
    def bindContext(self,serv:Optional[ServerContext],clie:Optional['RpcExtendClient1'])->'TableSerializer':
        self.boundServContext=serv
        self.boundClieContext=clie
        return self
    
    def bindSerializer(self,ser:Serializer)->'TableSerializer':
        self.ser=ser
        return self
    
    def getRowsData(self,rowCnt:int)->List[List[Any]]:
        headerType=NotNone(self.headerType)
        colCnt=len(headerType)
        ser=NotNone(self.ser)
        rows=[]
        for i1 in range(rowCnt):
            rows.append([None]*colCnt)
        for i1 in range(colCnt):
            typ=headerType[i1]
            if typ=='i':
                for i2 in range(rowCnt):
                    rows[i2][i1]=ser.getInt()
            elif typ=='l':
                for i2 in range(rowCnt):
                    rows[i2][i1]=ser.getLong()
            elif typ=='f':
                for i2 in range(rowCnt):
                    rows[i2][i1]=ser.getFloat()
            elif typ=='d':
                for i2 in range(rowCnt):
                    rows[i2][i1]=ser.getDouble()
            elif typ=='b':
                for i2 in range(rowCnt):
                    rows[i2][i1]=ser.getBytes()
            elif typ=='s':
                for i2 in range(rowCnt):
                    rows[i2][i1]=ser.getString()
            elif typ=='c':
                for i2 in range(rowCnt):
                    rows[i2][i1]=ser.getVarint()!=0
            elif typ=='o':
                for i2 in range(rowCnt):
                    idx=ser.getInt()
                    if idx==-1:
                        rows[i2][i1]=None
                    else:
                        if self.boundServContext!=None:
                            rows[i2][i1]=self.boundServContext.getRef(idx).object
                        else:
                            rows[i2][i1]=RpcExtendClientObject(NotNone(self.boundClieContext),idx)
            else:
                raise IOError('Unknown Type')
        return rows

    def putRowsData(self,rows:List[List[Any]]):
        headerType=NotNone(self.headerType)
        colsCnt=len(headerType)
        rowCnt=len(rows)
        ser=NotNone(self.ser)
        for i1 in range(colsCnt):
            typ=headerType[i1]
            if typ=='i':
                for i2 in range(rowCnt):
                    ser.putInt(rows[i2][i1])
            elif typ=='l':
                for i2 in range(rowCnt):
                    ser.putLong(rows[i2][i1])
            elif typ=='f':
                for i2 in range(rowCnt):
                    ser.putFloat(rows[i2][i1])
            elif typ=='d':
                for i2 in range(rowCnt):
                    ser.putDouble(rows[i2][i1])
            elif typ=='b':
                for i2 in range(rowCnt):
                    ser.putBytes(rows[i2][i1])
            elif typ=='s':
                for i2 in range(rowCnt):
                    ser.putString(rows[i2][i1])
            elif typ=='c':
                for i2 in range(rowCnt):
                    ser.putVarint(1 if rows[i2][i1] else 0)
            elif typ=='o':
                for i2 in range(rowCnt):
                    t2=rows[i2][i1]
                    if t2==None:
                        ser.putInt(-1)
                    else:
                        if self.boundServContext!=None:
                            ref2=allocRefFor(self.boundServContext,t2)
                            ser.putInt(ref2.index)
                        else:
                            ser.putInt(NotNone(cast(RpcExtendClientObject,t2).value))
            else:
                raise IOError('Unknown Type')

    def load(self,buf:bytes):
        if buf!=None:
            self.bindSerializer(Serializer().prepareUnserializing(buf))
        ser=NotNone(self.ser)
        flag=ser.getVarint()
        rowCnt=ser.getVarint()
        headerType=ser.getString()
        self.headerType=headerType
        colCnt=len(headerType)
        if (flag & TableSerializer.FLAG_NO_HEADER_NAME)==0:
            headerName2 = []
            for i1 in range(colCnt):
                headerName2.append(ser.getString())
            self.headerName=headerName2
        self.rows=self.getRowsData(rowCnt)
        return self

    def build(self):
        if self.ser==None:
            self.ser=Serializer().prepareSerializing()
        if self.headerType==None:
            if len(self.rows)>=1:
                self.headerType=''
                for t1 in self.rows[0]:
                    self.headerType+=pytypeToDeclChar(type(t1))
            else:
                self.headerType=''
        colsCnt=len(self.headerType)
        flag=0
        if self.headerName==None:
            flag|=TableSerializer.FLAG_NO_HEADER_NAME
        
        self.ser.putVarint(flag)
        rowCnt=len(self.rows)
        self.ser.putVarint(rowCnt)
        self.ser.putString(self.headerType)
        if self.headerName!=None:
            for e in self.headerName:
                self.ser.putString(e)
        self.putRowsData(self.rows)
        return self.ser.build()
    
    def toMapArray(self)->List[Dict[str,Any]]:
        r=[]
        assert self.headerName!=None
        for t1 in range(self.getRowCount()):
            r0={}
            row=self.getRow(t1)
            for t2,t3 in enumerate(self.headerName):
                r0[t3]=row[t2]
            r.append(r0)
        return r

    def fromMapArray(self,val:List[Dict[str,Any]]):
        if len(val)>=1 and self.headerName==None:
            self.headerName=list(val[0].keys())
        for t1 in val:
            row=[]
            for t2 in NotNone(self.headerName):
                row.append(t1[t2])
            self.addRow(row)
        return self
    
    def toArray(self)->List[Dict[str,Any]]:
        r=[]
        assert self.headerName!=None
        for t1 in range(self.getRowCount()):
            r.append(self.getRow(t1)[0])
        return r

    def fromArray(self,val:List[Dict[str,Any]]):
        if len(val)>=1 and self.headerName==None:
            self.headerName=list(val[0].keys())
        for t1 in val:
            self.addRow([t1])
        return self
    
    


class RpcExtendClientObject():
    def __init__(self,client:'RpcExtendClient1',value:Optional[int]=None):
        self.value=value
        self.client=client

    async def free(self):
        if self.value is not None:
            val=self.value
            self.value=None
            sid=self.client.allocSid()
            try:
                await self.client.conn.freeRef([val],sid)
            finally:
                self.client.freeSid(sid)
            

    async def asCallable(self):
        '''this object will be invalid after this function. use return value instead.'''
        c1=RpcExtendClientCallable(self.client,self.value)
        self.value=None
        return c1
            
    def __del__(self):
        if self.client.conn.running:
            asyncio.create_task(self.free())



class RpcExtendClientCallable(RpcExtendClientObject):
        
        
    def typedecl(self,decl:str):
        ''' function type declaration
format: 'parameters type->return type' 
eg:
a function defined in c:
    bool fn(uin32_t,uint64_t,float64_t,struct pxprpc_object *)
defined in java:
    boolean fn(int,int,double,Object)
    ...
it's pxprpc typedecl: 
    iido->z

available type typedecl characters:
  i  int(32bit integer)
  l  long(64bit integer)
  f  float(32bit float)
  d  double(64bit float)
  o  object(32bit reference address)
  b  bytes(bytes buffer)

  c  boolean(pxprpc use 1byte(1/0) to store a boolean value)
  s  string(bytes will be decode to string)
        '''
        self.tParam,self.tResult=decl.split('->')
        return self


    async def __call__(self,*args)->Any:
        assert self.value is not None
        if self.tParam=='b':
            packed=args[0]
        else:
            ser=Serializer().prepareSerializing()
            TableSerializer()\
                .bindContext(None,self.client).bindSerializer(ser).setHeader(self.tParam,None)\
                .putRowsData([list(args)])

            packed=ser.build()
        sid=self.client.allocSid()
        try:
            result=await self.client.conn.call(self.value,packed,sid)
            if self.tResult=='b':
                return result
            else:
                ser=Serializer().prepareUnserializing(result)
                results=TableSerializer()\
                    .bindContext(None,self.client).bindSerializer(ser).setHeader(self.tResult,None)\
                    .getRowsData(1)[0]

                if len(self.tResult)==0:
                    return None
                elif len(self.tResult)==1:
                    return results[0]
                else:
                    return results
        finally:
            self.client.freeSid(sid)




class RpcExtendError(Exception):
    pass

class RpcExtendClient1:
    def __init__(self,conn:ClientContext):
        self.conn=conn
        self.__usedSid:Set[int]=set()
        self.__sidStart=1
        self.__sidEnd=0xffff
        self.__nextSid=self.__sidStart
        self.builtIn=None
        self.serverName=None

    async def init(self):
        asyncio.create_task(self.conn.run())
        info=await self.conn.getInfo()
        for t1 in info.split('\n'):
            if ':' in t1:
                key,value=t1.split(':')
                if key=='server name':
                    self.serverName=value

    def allocSid(self)->int:
        reachEnd=False
        while self.__nextSid in self.__usedSid:
            self.__nextSid+=1
            if self.__nextSid>=self.__sidEnd:
                if reachEnd:
                    raise RpcExtendError('No sid available')
                else:
                    reachEnd=True
                    self.__nextSid=self.__sidStart

        t1=self.__nextSid
        self.__nextSid+=1
        if self.__nextSid>=self.__sidEnd:
            self.__nextSid=self.__sidStart
        self.__usedSid.add(t1)
        return t1

    def freeSid(self,index:int):
        try:
            self.__usedSid.remove(index)
        except KeyError:
            pass

    async def getFunc(self,name:str)->Optional[RpcExtendClientCallable]:
        sid=self.allocSid()
        try:
            index=await self.conn.getFunc(name,sid)
            if index==-1:
                return None
            return RpcExtendClientCallable(self,value=index)
        finally:
            self.freeSid(sid)


def allocRefFor(context:ServerContext,obj:Any)->PxpRef:
    ref=context.allocRef()
    ref.object=obj
    if hasattr(obj,'close'):
        def fn():
            obj.close()
        ref.onFree=fn
    return ref

class decorator:
    @staticmethod
    def typedecl(decl):
         def fn2(fn):
            fn._pxprpc__PyCallableWrapTypedecl=decl
            return fn
         return fn2


class PyCallableWrap(PxpCallable):
    
    def __init__(self,c:Callable,classMethod:bool):
        self.tParam=''
        self.tResult=''
        self.callable=c
        if hasattr(c,'_pxprpc__PyCallableWrapTypedecl'):
            self.typedecl(c._pxprpc__PyCallableWrapTypedecl)
        else:
            argsInfo:inspect.FullArgSpec=inspect.getfullargspec(c)
            if hasattr(argsInfo,'args'):
                for argName in (argsInfo.args[1:] if classMethod else argsInfo.args):
                    t1=argsInfo.annotations[argName]
                    self.tParam+=pytypeToDeclChar(t1)
                    
            if 'return' in argsInfo.annotations:
                t1=argsInfo.annotations['return']
                self.tResult+=pytypeToDeclChar(t1)
            
    def typedecl(self,decl:str)->'PyCallableWrap':
        self.tParam,self.tResult=decl.split('->')
        return self
    
    def readParameter(self,req:PxpRequest)->List[Any]:
        assert req.context!=None
        para=[]
        if self.tParam=='b':
            para.append(req.parameter)
        else:
            ser=Serializer().prepareUnserializing(req.parameter)
            para=TableSerializer()\
                .bindContext(req.context,None).bindSerializer(ser).setHeader(self.tParam,None)\
                .getRowsData(1)[0]
        return para
    
    async def call(self,req:PxpRequest):
        para=self.readParameter(req)
        try:
            r=await self.callable(*para)
            req.result=self.writeResult(req,r)
        except Exception as ex:
            req.rejected=ex
    
    def writeResult(self,req:PxpRequest,result:Any)->bytes:
        assert req.context!=None
        if self.tResult=='b':
            return cast(bytes,result)
        else:
            ser=Serializer().prepareSerializing()
            result2=[result] if len(self.tResult)<=1 else result
            TableSerializer()\
                .bindContext(req.context,None).bindSerializer(ser).setHeader(self.tResult,None)\
                .putRowsData([result2])
            b=ser.build()
            return b
        
RegisteredFuncMap:Dict[str,Any]=dict()

def CDefaultExtendFuncMap(name:str)->Optional[PxpCallable]:
    t2=RegisteredFuncMap.get(name,None)
    if t2!=None:
        if isinstance(t2,PxpCallable):
            return t2
        else:
            return PyCallableWrap(t2,False)
    
    delimIndex=name.rfind('.')
    if delimIndex>=0:
        t2=RegisteredFuncMap.get(name[0:delimIndex],None)
        subfuncName=name[delimIndex+1:]
        if t2!=None and hasattr(t2,subfuncName):
            t1=getattr(t2,subfuncName)
            if t1!=None:
                return PyCallableWrap(t1,True)
    return None



class builtinFuncs:
    def __init__(self):
        self.pxprpcGlobals={}
    async def anyToString(self,obj:object):
        return str(obj)

    async def pyExec(self,code:str):
        exec(code,self.pxprpcGlobals,locals())

    async def pyGlobalGet(self,name:str)->Any:
        return self.pxprpcGlobals[name]
    
    async def pyAwait(self,awaitable:Awaitable):
        await awaitable

    async def bufferData(self,buf:Any)->bytes:
        return buf

RegisteredFuncMap['builtin']=builtinFuncs()

from . import base
base.DefaultFuncMap=CDefaultExtendFuncMap